### ISSUE TITLE
Slowness with cancancan 3.2+

### ISSUE THREAD

**Issue by mathieujobin** [CanCanCommunity/cancancan#716]
### Steps to reproduce

I have a cucumber scenario that takes 17 seconds when using cancancan 3.1.0
and it went up to 2m40 after upgrading cancancan.

We do make extensive use of cancancan. Right now I don't know the exact cause, I just rollback to fix the emergency. But I will post more details when I find out more.

quick compare link for the change log https://github.com/CanCanCommunity/cancancan/compare/3.1.0..3.2.0?w=1

-----
Please copy the following Gist and edit it to reproduce your issue: https://gist.github.com/coorasse/3f00f536563249125a37e15a1652648c

### Expected behavior
My scenario used to run fast

### Actual behavior
Its slow, i confirm with both 3.2 and 3.3, its slow
with 3.1 its fast

### System configuration
**Rails version**:
5.2.6

**Ruby version**:
2.6.7

**CanCanCan version**
3.1 versus 3.3

**Comment by mathieujobin** [comment]
Added a timeout in my code such as this

```
     def can_access_url?(url, options = {})
+out = nil
+calculation_time = Benchmark.realtime do
+Timeout.timeout(10) do
+     out = foo_can_access_url?(url, options)
+    end
+    end
+warn "CANCAN: #{calculation_time}"
+out
+    rescue
+debugger
+puts "what"
+    end
+    def foo_can_access_url?(url, options = {})

```

stacktrace was:

```
 ["/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/class_matcher.rb:15:in `matching_class_check'",
 "/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/class_matcher.rb:8:in `block in matches_subject_class?'",
 "/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/class_matcher.rb:6:in `any?'",
 "/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/class_matcher.rb:6:in `matches_subject_class?'",
 "/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/rule.rb:114:in `matches_subject_class?'",
 "/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/rule.rb:110:in `matches_subject?'",
 "/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/ability/rules.rb:57:in `relevant_rules_for_match'",
 "/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/ability.rb:76:in `block in can?'",
 "/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/ability.rb:79:in `each'",
 "/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/ability.rb:79:in `each'",
 "/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/ability.rb:79:in `each'",
 "/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/ability.rb:79:in `each'",
 "/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/ability.rb:79:in `first'",
 "/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/ability.rb:79:in `can?'",
 "/home/mathieu/.rvm/rubies/ruby-2.6.6/lib/ruby/gems/2.6.0/gems/cancancan-3.2.0/lib/cancan/controller_additions.rb:377:in `can?'"]
```

staytuned for more

**Comment by mathieujobin** [comment]
I have pinpointed to SubjectClassMatcher being slow. I don't know if it can be disabled?

**Comment by mathieujobin** [comment]
![image](https://user-images.githubusercontent.com/99191/123577517-f23a9780-d80e-11eb-935e-0eec27a116cd.png)

it appears calling `subclasses` is most expensive.

https://github.com/CanCanCommunity/cancancan/compare/3.1.0..3.2.0?w=1#diff-8b4268895f6f86a20348eb1e6d21006d32e01be8acac375a2bf0578b29f38861R15

Maybe we could memoize the result of these methods in case the same subjects are requested multiple times. in my case, it shows 303 calls to ObjectSpace#each_object

![object-space-cancan](https://user-images.githubusercontent.com/99191/123578117-164aa880-d810-11eb-80cb-861ca0be5682.png)

**Comment by coorasse** [comment]
@Liberatys any input on this issue? @mathieujobin could you somehow share some relevant code? I see and acknowledge that this is making your code slower but...wow! I'd be interested in checking why is it called so often? 
With cancancan 3.2.0 we have a configuration, so we might also introduce the possibility to disable STI (that's why `descendats` is used here I guess) but before allowing that, I'd like to have more context and exclude that the issue is possibly somewhere else, specifically in **how** cancancan is used here.

**Comment by Liberatys** [comment]
@coorasse ... Sorry for not responding 🤦🏻 I will take care of some STI improvements over the comming weeks 👍🏻

**Comment by mathieujobin** [comment]
@coorasse we're calling cancancan for every links to be displayed, that's probably what exacerbate the problem for us.
I don't think there is any other relevant code to share 

we have the gem locked at 3.1 until this issue is fixed.

thank you both for looking into

cc: @Liberatys

**Comment by Liberatys** [comment]
@mathieujobin hey there ^^ Sorry that I never write in this issue directly 👍🏻 Some time ago we merge some prs that should have improved your experience by caching the subclasses during the evaluation of the rules. Please have a try 👍🏻.

**Comment by mathieujobin** [comment]
Can you link to a specific one?

**Comment by mathieujobin** [comment]
@Liberatys I confirm, the slowness is present on 3.2.2, 3.3.0 but gone with 9c75a78b5b60d59d9620 `develop`  branch

Thank you

**Comment by coorasse** [comment]
This is now released as 3.4.0
